test/module/scheduler_test.cpp

```diff
@@ -293,6 +293,63 @@ TEST_F(SchedulerTest, SimpleMultiHomeBatch) {
   ASSERT_EQ(output_txn.write_set().at("Z"), "newZ");
 }
 
+TEST_F(SchedulerTest, SinglePartitionTransactionValidateMasters) {
+  auto txn = MakeTransaction(
+      {"A"}, /* read_set */
+      {"D"},  /* write_set */
+      "GET A     \n"
+      "SET D newD\n", /* code */
+      {{"A", {0,0}}, {"D", {0,0}}},
+      MakeMachineId("0:1") /* coordinating server */);
+  txn->mutable_internal()->set_type(TransactionType::SINGLE_HOME);
+
+  SendSingleHomeBatch(100, {txn}, {0});
+
+  auto output_txn = ReceiveMultipleAndMerge(1, 1);
+  LOG(INFO) << output_txn;
+  ASSERT_EQ(output_txn.status(), TransactionStatus::COMMITTED);
+  ASSERT_EQ(output_txn.read_set_size(), 1);
+  ASSERT_EQ(output_txn.read_set().at("A"), "valueA");
+  ASSERT_EQ(output_txn.write_set_size(), 1);
+  ASSERT_EQ(output_txn.write_set().at("D"), "newD");
+}
+
+TEST_F(SchedulerTest, SinglePartitionTransactionProcessRemaster) {
+  auto txn = MakeTransaction(
+      {"A"}, /* read_set */
+      {},  /* write_set */
+      "GET A     \n", /* code */
+      {{"A", {1,1}}},
+      MakeMachineId("0:0") /* coordinating server */);
+  txn->mutable_internal()->set_type(TransactionType::SINGLE_HOME);
+  txn->mutable_internal()->set_id(10);
+
+  auto remaster_txn = MakeTransaction(
+      {}, /* read_set */
+      {"A"},  /* write_set */
+      "", /* code */
+      {{"A", {0,0}}}, /* master metadata */
+      MakeMachineId("0:1") /* coordinating server */,
+      1 /* new master */);
+  remaster_txn->mutable_internal()->set_type(TransactionType::SINGLE_HOME);
+  remaster_txn->mutable_internal()->set_id(11);
+
+  SendSingleHomeBatch(100, {txn, remaster_txn}, {0});
+
+  auto output_remaster_txn = ReceiveMultipleAndMerge(1, 1);
+  LOG(INFO) << output_remaster_txn;
+  ASSERT_EQ(output_remaster_txn.internal().id(), 11);
+  ASSERT_EQ(output_remaster_txn.status(), TransactionStatus::COMMITTED);
+  ASSERT_EQ(output_remaster_txn.new_master(), 1);
+
+  auto output_txn = ReceiveMultipleAndMerge(0, 1);
+  LOG(INFO) << output_txn;
+  ASSERT_EQ(output_txn.internal().id(), 10);
+  ASSERT_EQ(output_txn.status(), TransactionStatus::COMMITTED);
+  ASSERT_EQ(output_txn.read_set_size(), 1);
+  ASSERT_EQ(output_txn.read_set().at("A"), "valueA");
```

Comment:
The transactions may be returned out of order. You should look at the IDs to decide which one is which.

Reply:
Oh. I just realized you send them to two different servers.